src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -30,19 +32,429 @@
+
+            Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentText((DocumentId)null, text, PreservationMode.PreserveIdentity));
+            Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentText(s_documentId, text, PreservationMode.PreserveIdentity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithAnalyzerConfigDocumentText_TextAndVersion()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().State.AnalyzerConfigDocumentStates.Single().Key;
+            var textAndVersion = TextAndVersion.Create(SourceText.From("new text"), VersionStamp.Default);
+
+            var newSolution1 = solution.WithAnalyzerConfigDocumentText(documentId, textAndVersion, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetAnalyzerConfigDocument(documentId).TryGetText(out var actualText));
+            Assert.True(newSolution1.GetAnalyzerConfigDocument(documentId).TryGetTextVersion(out var actualVersion));
+            Assert.Equal(textAndVersion.Text, actualText);
+            Assert.Equal(textAndVersion.Version, actualVersion);
+
+            var newSolution2 = newSolution1.WithAnalyzerConfigDocumentText(documentId, textAndVersion, PreservationMode.PreserveIdentity);
+            Assert.Same(newSolution1, newSolution2);
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithAnalyzerConfigDocumentText(documentId, (SourceText)null, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithAnalyzerConfigDocumentText(documentId, textAndVersion, (PreservationMode)(-1)));
+
+            Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentText((DocumentId)null, textAndVersion, PreservationMode.PreserveIdentity));
+            Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentText(s_documentId, textAndVersion, PreservationMode.PreserveIdentity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentTextLoader()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var loader = new TestTextLoader("new text");
+
+            var newSolution1 = solution.WithDocumentTextLoader(documentId, loader, PreservationMode.PreserveIdentity);
+            Assert.Equal("new text", newSolution1.GetDocument(documentId).GetTextSynchronously(CancellationToken.None).ToString());
+
+            // Reusal is not currently implemented:
+            var newSolution2 = solution.WithDocumentTextLoader(documentId, loader, PreservationMode.PreserveIdentity);
+            Assert.NotSame(newSolution1, newSolution2);
```

Comment:
Is asserting a non-optimal behavior worth it?

Reply:
Yes. Tests should capture current behavior.

Reply:
@tmat Eh, might be worth adding as stronger comment that the behavior *can* be changed. Sometimes somebody might fix this bug by accident, see it's breaking tests, and then figure they can't make the change even though it was a good thing.

Reply:
I think having the linked issue is good enough indicator.